lib/src/test/java/com/auth0/jwt/JWTCreatorTest.java

```diff
@@ -717,4 +717,76 @@ public void shouldRefuseCustomListClaimForUnknownArrayType() throws Exception {
                 .sign(Algorithm.HMAC256("secret"));
     }
 
+    @SuppressWarnings("Convert2Diamond")
+    @Test
+    public void shouldAddPayloadClaim() throws Exception {
+        Map<String, Object> payload = new HashMap<String, Object>();
+        payload.put("asd", 123);
+        String signed = JWTCreator.init().withPayload(payload).sign(Algorithm.HMAC256("secret"));
+
+        assertThat(signed, is(notNullValue()));
+        String[] parts = signed.split("\\.");
+        String payloadJson = new String(Base64.decodeBase64(parts[1]), StandardCharsets.UTF_8);
+        assertThat(payloadJson, JsonMatcher.hasEntry("asd", 123));
+    }
+
+    @Test
+    public void shouldReturnBuilderIfNullMapIsProvidedForPayload() throws Exception {
+        String signed = JWTCreator.init().withPayload(null).sign(Algorithm.HMAC256("secret"));
+
+        assertThat(signed, is(notNullValue()));
+    }
+
+    @Test
+    public void shouldOverwriteExistingPayloadIfPayloadMapContainsTheSameKey() throws Exception {
+        Map<String, Object> payload = new HashMap<String, Object>();
+        payload.put(PublicClaims.KEY_ID, "xyz");
+
+        String signed = JWTCreator.init().withKeyId("abc").withPayload(payload).sign(Algorithm.HMAC256("secret"));
+
+        assertThat(signed, is(notNullValue()));
+        String[] parts = signed.split("\\.");
+        String payloadJson = new String(Base64.decodeBase64(parts[1]), StandardCharsets.UTF_8);
+        assertThat(payloadJson, JsonMatcher.hasEntry(PublicClaims.KEY_ID, "xyz"));
+    }
+
+    @Test
+    public void shouldOverwriteExistingPayloadsWhenSettingSamePayloadKey() throws Exception {
+        Map<String, Object> payload = new HashMap<String, Object>();
+        payload.put(PublicClaims.ISSUER, "xyz");
+
+        String signed = JWTCreator.init().withPayload(payload).withIssuer("abc").sign(Algorithm.HMAC256("secret"));
+
+        assertThat(signed, is(notNullValue()));
+        String[] parts = signed.split("\\.");
+        String payloadJson = new String(Base64.decodeBase64(parts[1]), StandardCharsets.UTF_8);
+        assertThat(payloadJson, JsonMatcher.hasEntry(PublicClaims.ISSUER, "abc"));
+    }
+
+    @Test
+    public void shouldRemovePayloadIfTheValueIsNull() throws Exception {
+        Map<String, Object> payload = new HashMap<String, Object>();
+        payload.put(PublicClaims.KEY_ID, null);
+        payload.put("test2", "isSet");
+
+        String signed = JWTCreator.init().withKeyId("test").withPayload(payload).sign(Algorithm.HMAC256("secret"));
```

Comment:
This test would fail as implemented, as the claim validation of map entries does not allow nulls. We probably want to maintain that consistency as currently implemented for lists and maps.